libdasd: add common dasd ioctls to the lib

r4f4 opened the pull request:
This avoids code duplication on dasd-related programs.

Signed-off-by: Rafael Fonseca <r4f4rfs@gmail.com>

michael-holzheu commented:
@hoeppnerj that's for you.

r4f4 commented on include/lib/dasd_base.h:
```diff
@@ -0,0 +1,243 @@
+        char characteristics[64];   /* from read_device_characteristics */
+        char configuration_data[256]; /* from read_configuration_data */
+} dasd_information_t;
+
```
Good idea. I'll do that.

r4f4 replied:
I have one question though: in dasdview, extra info is only printed in case dasd_info_version >= 2. How would you detect the info version when using only BIODASDINFO2?

hoeppnerj replied:
That check can be dropped as well. BIODASDINFO2 always provides all information. Both IOCTLs (BIODASDINFO and BIODASDINFO2) call the exact same function within the kernel. The only difference is, that BIODASDINFO copies less data (Have a look at dasd_ioctl_information() in drivers/s390/block/dasd_ioctl.c in the [kernel](https://github.com/torvalds/linux/blob/4fbd8d194f06c8a3fd2af1ce560ddb31f7ec8323/drivers/s390/block/dasd_ioctl.c#L408))

So, it's totally fine to drop checks in that regard.

r4f4 commented on libdasd/dasd_ioctl.c:
```diff
@@ -0,0 +1,364 @@
+	warn("Could not open device '%s'", device)
+
+int dasd_get_blocksize_from_fd(int fd, unsigned int *blksize)
+{
```
I added the _from_fd functions because in some cases the device is already open in the app using the lib. I have no problem in dropping them though.

hoeppnerj replied:
Yea, I know. I was debating that part a bit with myself, but I think it's for the better to have one set of functions that are used consistently across all tools rather than different variants. I also think the better attempt is to provide a minimalistic library (as in as little function variants as possible) and then adapt the tools accordingly. The library should not be geared to the tools. It should be the other way around.

r4f4 commented on libdasd/dasd_ioctl.c:
```diff
@@ -0,0 +1,364 @@
+		COULD_NOT_OPEN_WARN(device);
+		return errno;
+	}
+
```
If we exit from lib, we are not giving the app a chance to run cleanup routines. Another problem might arise when the ioctl is "optional" but we'd still exit the app because the lib makes all of them fatal. That's why I tried to give the app a chance to decide on what to do in case the ioctl failed.

hoeppnerj replied:
That was, or rather still is, one of my concerns as well. But I wanted to discuss it anyway. My idea was a certain level of abstraction. From a user's point of view, I don't want to care about IOCTLs. I just want to call a function that gives me certain informations. The library should deal with the rest.

But you're completely right, there is no chance for the app to do cleanups. I don't really have a better idea at the moment to combine the abstraction idea with a level of safety for the app. Hm..

You can leave your approach as it is for now.

r4f4 commented on libdasd/dasd_ioctl.c:
```diff
@@ -0,0 +1,364 @@
+ * @retval errno	in case of failure
+ */
+int dasd_get_blocksize64(const char *device, unsigned long long *blksize)
+{
```
Ack.

r4f4 commented on libdasd/dasd_ioctl.c:
```diff
@@ -0,0 +1,364 @@
+ * @retval 0		in case of success
+ * @retval errno	in case of failure
+ */
+int dasd_get_ro(const char *device, int *ro)
```
The implementation of this function is related to my comment above about making the functions in this lib non-fatal. If you think this format is more useful, I can change it.

r4f4 commented on libdasd/dasd_ioctl.c:
```diff
@@ -0,0 +1,364 @@
+ * @retval 0		in case of success
+ * @retval errno	in case of failure
+ */
+int dasd_disk_format(const char *device, format_data_t *p)
```
Ack.

r4f4 commented on libdasd/dasd_ioctl.c:
```diff
@@ -0,0 +1,364 @@
+ * @retval 0		in case of success
+ * @retval errno	in case of failure
+ */
+int dasd_format_check(const char *device, format_check_t *p)
```
Ack.

r4f4 commented:
Sure, I can divide the patch into several smaller commits. I'll do that for the next version.

hoeppnerj commented on dasdfmt/dasdfmt.c:
```diff
@@ -1071,81 +1051,99 @@ static void dasdfmt_write_labels(dasdfmt_info_t *info, volume_label_t *vlabel,
 		vlabel->formatted_blocks = cylinders * heads * geo.sectors;
-		rc = write(filedes, &vlabel->vollbl, (sizeof(*vlabel)
+		rc = write(fd, &vlabel->vollbl, (sizeof(*vlabel)
 						     - sizeof(vlabel->volkey)));
```
Same here

hoeppnerj commented on dasdfmt/dasdfmt.c:
```diff
@@ -1071,81 +1051,99 @@ static void dasdfmt_write_labels(dasdfmt_info_t *info, volume_label_t *vlabel,
 	if (info->cdl_format) {
-		rc = write(filedes, vlabel, (sizeof(*vlabel) -
+		rc = write(fd, vlabel, (sizeof(*vlabel) -
 					     sizeof(vlabel->formatted_blocks)));
```
Little indentation error here after renaming filedes to fd

r4f4 commented:
@hoeppnerj Hopefully I addressed all your raised points. I also added 2 more commits fixing the coding style in dasdinfo and dasdview.
Let me know if there is anything else.

michael-holzheu commented on include/lib/dasd_base.h:
```diff
@@ -0,0 +1,207 @@
+ *
+ * Copyright IBM Corp. 2013, 2017
+ * Copyright RedHat Inc. 2017
+ *
```
Wow, the first "extraterrestrial" Copyright notice in s390-tools ;-)

Also in other projects like util-linux or the linux kernel multiple Copyright statements are added to single files - so this is common practice:

```
~/linux/kernel $ git grep Copyright
auditsc.c: * Copyright 2003-2004 Red Hat Inc., Durham, North Carolina.
auditsc.c: * Copyright 2005 Hewlett-Packard Development Company, L.P.
auditsc.c: * Copyright (C) 2005, 2006 IBM Corporation
auditsc.c: * okir@suse.de and Copyright 2003 SuSE Linux AG.
bpf/arraymap.c:/* Copyright (c) 2011-2014 PLUMgrid, http://plumgrid.com
bpf/arraymap.c: * Copyright (c) 2016,2017 Facebook

~/util-linux $ git grep Copyright
disk-utils/fdisk.c: * Copyright (C) 1992  A. V. Le Blanc (LeBlanc@mcc.ac.uk)
disk-utils/fdisk.c: * Copyright (C) 2012  Davidlohr Bueso <dave@gnu.org>
disk-utils/fdisk.c: * Copyright (C) 2007-2013 Karel Zak <kzak@redhat.com>
```

But when this should be done? How do we handle it when we again move code to other places and everything is mixed up? I would like to define at least some rough rules how we handle this for s390-tools.

Since this is the only file where you added the Copyright statement, I assume you followed some rules practiced by Red Hat. So what would be the Red Hat way to handle this?

BTW: What's the right spelling for the Copyright line, "RedHat" or "Red Hat"?

r4f4 replied:
I only added to this file because I started if from scratch and this was a placeholder I put which I should've gone back to and fixed. The correct spelling is Red Hat. I'll fix that and the other files with the proper copyright lines.

michael-holzheu replied:
>  I'll fix that and the other files with the proper copyright lines.

Do we really want to add a Copyright notice for every single change?

r4f4 replied:
You are right, that's not needed. Most of the commits are just removing code. Since the most significant change is on libdasd, I guess that's the one which should have the copyright notice.

michael-holzheu replied:
>  Since the most significant change is on libdasd, I guess that's the one which should have the copyright notice.

I just talked to our legal expert: The "rule" is that you can add your copyright statement for "not trivial substantial" contributions. What the "substantial" means will be decided by the s390-tools developers and maintainers.

So I would say the copyright for `dasd_ioctl.c` is ok. The copyright for `dasd_base.h` might be debatable, but I would also accept that one.

hoeppnerj commented:
@r4f4 I just found a major issue that I didn't pay attention to or rather didn't remember while reviewing.

Dasdfmt currently doesn't work since we're closing the file descriptor after each IOCTL call. Before BIODASDFMT can be called, a DASD has to be disabled (or rather put into a BASIC state) via BIODASDDISABLE. However, if you disable the DASD, close the file descriptor, and then try to reopen it, it won't work as the device isn't fully usable anymore. For BIODASDFMT to work, the file descriptor opened for BIODASDISABLE has to be kept open until BIODASDFMT has finished.

One approach now would be, as BIODASD{DISABLE,ENABLE} and BIODASDFMT only make sense together, to put the calls for BIODASDDISABLE and *ENABLE into the dasd_format_disk() function and not expose dasd_disk_{enable,disable} at all. However, then it gets a bit fiddly to figure out the state of the disk in case the program gets interrupted and we want to re-enable the disk.

Therefore, my suggestion now would be to have dasd_disk_disable() return the file descriptor and change dasd_format_disk() to take an fd as an additional parameter (You already had this approach in the first iteration were you could pass a file description to some functions). I think thats less effort for now and the other approach could be worked on at a later time.

Sorry for the inconveniences, I just noticed it while doing some basic function tests with your patches. (Should've tested it earlier really).

r4f4 commented:
@hoeppnerj thank you for testing. I tested some of the tools, but I couldn't get hold of a s390x machine where I could run the formatting commands. I'll fix that and update the patches.

michael-holzheu commented:
> but I couldn't get hold of a s390x machine where I could run the formatting commands.

@r4f4 Red Hat has a Mainframe with attached DASDs for testing. IMHO it is *important* that you organize a test setup when you write DASD related s390-tools code.

r4f4 commented:
dasdfmt is working now but I really dislike the way dasd_(dis|en)able and dasd_format_disk don't conform to the same structure as the other functions.

What if we made dasd_ioctl.c context-based? dasd_ioctl_init() could initialize a ctx struct containing some info (like file descriptor and disk state) and all operations would use that context on an open device. We would only close the device upon dasd_ioctl_deinit();

hoeppnerj commented:
> dasdfmt is working now but I really dislike the way dasd_(dis|en)able and dasd_format_disk don't conform to the same structure as the other functions.
> 
> What if we made dasd_ioctl.c context-based? dasd_ioctl_init() could initialize a ctx struct containing some info (like file descriptor and disk state) and all operations would use that context on an open device. We would only close the device upon dasd_ioctl_deinit();

I'm not entirely happy with it as well, but I think the solution is fine for now. A better way to handle this issue can be worked on at a later point. I'm not sure about the context-based either. I appreciate the suggestion though and will take it with me to our internal discussion.

A few more review comments following in a bit.

hoeppnerj commented on dasdfmt/dasdfmt.c:
```diff
@@ -415,6 +400,30 @@ static void evaluate_format_error(dasdfmt_info_t *info, format_check_t *cdata,
+	disk_disabled = 0;
+}
+
+static void disk_disable(const char *device, int *fd)
```
You're passing the global file descriptor to these functions all the time. I would suggest to either use the global file descriptor here directly or use local fds in the respective places. (I would prefer the latter)

r4f4 replied:
Ok. I'll submit a fixed version soon.

r4f4 replied:
Oh, I can't quite get rid of the global `filedes` because the callback for signal interruption needs it to re-access the device. So I can change `process_tracks` to receive a file descriptor. However, for `disk_disable` and `disk_enable`, since they already write to the global var `disk_disabled`, what if they also wrote/used directly `filedes`? I think that's better than doing all the time:
```
filedes = disk_disable(device);
[...]
disk_enable(filedes);
```
In that way, it would be clear that disk_enable|disable are wrappers to dasd_disk_enable|disable which also keep some global states.

hoeppnerj replied:
Oh, you're right. Missed that.
Yes, then the first suggestion, use the filedes directly in disk_(dis|en)able

hoeppnerj commented:
Alright, looks fine now. I merged and pushed it to our devel branch.
It'll be put upstream with the next sync.
Thanks @r4f4 for the effort!

r4f4 commented:
@hoeppnerj Any idea when the next sync is going to happen? I'd like to rebase my zipl_helper branch onto  master with the libdasd changes so that I can submit a new patch version.

stefan-haberland commented:
I will do the sync within this week as soon as possible.
